### PR TITLE
Gradle dependency checksum and signature checks

### PR DESCRIPTION
**Pull request by AndyScherzinger**
Adding checksums to Android repos like `android-common` for gradle based build toolchain, see https://docs.gradle.org/current/userguide/dependency_verification.html.

In order to do so, one must:

* add `org.gradle.dependency.verification.console=verbose` to `gradle.properties` so future CI runs would provide the info in case the HTML report isn't accessible
* run in the project root
  * `./gradlew --write-verification-metadata pgp,sha256 help`
  * `./gradlew :sample:generateDebugLintReportModel --write-verification-metadata pgp,sha256 help`
  * `./gradlew :core:spotlessKotlinCheck --write-verification-metadata pgp,sha256 help`
  * `./gradlew --export-keys`
  * `./gradlew :sample:generateDebugLintReportModel --export-keys`
  * `./gradlew :core:spotlessKotlinCheck --export-keys`
* ensure you commit `verification-keyring.keys` but not the binary very so git is able to diff the key changes. not executing the last command and commiting its results leads to the GH action trying to download the keys, this slows down the execution and leads to red checks since it'll fail for many keys, hence shipping them in the repo it the best option.

A next step will have to be to move to renovatebot which is capable of update checksums when updating dependencies, see https://docs.renovatebot.com/modules/manager/gradle/

---

## Note

* Due to poor signature management in the library world, many libraries aren't signed or their keys aren't retrievable❗ 
* The above command bootstraps the whole checksum+signature information and "blindly" trusts the given key. In theory all key would need to be verified. So this setup mechanism will only secure dependency changes of the future - an already compromised dependency would not be detected/detectable this way. While keeping in mind that a legit signature isn't an absolute statement in terms of the package being legit. It just means built and published by somebody with access to the keys.

### CODE REVIEW

**Comment by stefan-niedermann** [comment]
Slightly Off Topic: I am currently using the [`gradle/wrapper-validation-action`](https://github.com/gradle/wrapper-validation-action/issues/161) GitHub action in all of my gradle repositories ([Example](https://github.com/stefan-niedermann/nextcloud-deck/blob/master/.github/workflows/android.yml#L9-L16)). It's for validating the gradle wrapper itself though, not for dependencies.

**Comment by AndyScherzinger** [comment]
@stefan-niedermann we do so too on Android  repos - like this one - https://github.com/nextcloud/android-common/blob/main/.github/workflows/gradle-wrapper-validation.yml

**Comment by stefan-niedermann** [comment]
Aah, it's a separate workflow, that's why I missed it 🙈

**Comment by AndyScherzinger** [comment]
Fun stuff....

getting the checks to turn green is a major pain - as expected since metadata quality on libraries / dependency management are rather poor, especially for Android / Google artifacts.

I'll keep working on it but might keep going for the approach to put them on a trusted list because there is no way for my to maintain checksums manually because they aren't available anywhere :(

File: gradle/verification-metadata.xml [hunk 1]
```diff
@@ -0,0 +1,3275 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<verification-metadata xmlns="https://schema.gradle.org/dependency-verification" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="https://schema.gradle.org/dependency-verification https://schema.gradle.org/dependency-verification/dependency-verification-1.2.xsd">
+   <configuration>
+      <verify-metadata>true</verify-metadata>
+      <verify-signatures>true</verify-signatures>
+      <trusted-artifacts>
```
Comment (AndyScherzinger) [inline gradle/verification-metadata.xml:6]: some artifacts, thanks Google can't be verified, so put them in a trusted list

File: gradle/verification-metadata.xml [hunk 2]
```diff
@@ -0,0 +1,3275 @@
...
+         <trust group="com.android.tools.lint" name="lint-api" version="31.2.0" reason="Broken: checksum is missing from verification metadata"/>
+         <trust group="com.android.tools.lint" name="lint-checks" version="31.2.0" reason="Broken: checksum is missing from verification metadata"/>
+         <trust group="com.android.tools.lint" name="lint-gradle" version="31.2.0" reason="Broken: checksum is missing from verification metadata"/>
+         <trust group="org.apache.httpcomponents" name="httpclient" version="4.5.6" reason="Broken: artifact was signed but all keys were ignored, checksum is missing from verification metadata"/>
+         <trust group="org.jetbrains.kotlin" name="kotlin-stdlib" version="1.9.0" reason="Broken: artifact was signed but all keys were ignored, checksum is missing from verification metadata"/>
+         <trust group="org.jetbrains.kotlin" name="kotlin-stdlib-common" version="1.9.0" reason="Broken: artifact was signed but all keys were ignored, checksum is missing from verification metadata"/>
+      </trusted-artifacts>
+      <ignored-keys>
+         <ignored-key id="a41f13c999945293" reason="Key couldn't be downloaded from any key server"/>
```
Comment (AndyScherzinger) [inline gradle/verification-metadata.xml:21]: some keys aren't available via any dependency repo, so they will get added to an ignore list automatically

File: gradle/verification-metadata.xml [hunk 3]
```diff
@@ -0,0 +1,3275 @@
...
+         <ignored-key id="0da8a5ec02d11ead" reason="Key couldn't be downloaded from any key server"/>
+         <ignored-key id="ab2da4527f6ffc0b" reason="Key couldn't be downloaded from any key server"/>
+         <ignored-key id="fd5dea07fcb690a8" reason="Key couldn't be downloaded from any key server"/>
+         <ignored-key id="8671a8df71296252" reason="Key couldn't be downloaded from any key server"/>
+         <ignored-key id="6a0975f8b1127b83" reason="Key couldn't be downloaded from any key server"/>
+         <ignored-key id="4eb27db2a3b88b8b" reason="Key couldn't be downloaded from any key server"/>
+         <ignored-key id="5f69ad087600b22c" reason="Key couldn't be downloaded from any key server"/>
+      </ignored-keys>
+      <trusted-keys>
+         <trusted-key id="015479e1055341431b4545ab72475fd306b9cab7" group="com.googlecode.javaewah" name="JavaEWAH" version="1.2.3"/>
```
Comment (AndyScherzinger) [inline gradle/verification-metadata.xml:58]: same for found keys, they will be added to a trusted list

**Pull request merged**